qa/L0_implicit_state/implicit_state.py

```diff
@@ -0,0 +1,135 @@
+        inputs[0].set_data_from_numpy(np.random.randint(5, size=[1], dtype=np.int32))
+        inputs[1].set_data_from_numpy(np.asarray([0], dtype=np.int32))
+
+        with self.assertRaises(InferenceServerException) as e:
+            triton_client.infer(model_name="wrong_internal_state", inputs=inputs, sequence_id=2, sequence_start=True)
+
+        self.assertEqual(str(e.exception), "state 'undefined_state' is not a valid state name.")
+
+    def test_no_update(self):
+	    # Test implicit state without updating any state
+        triton_client = tritonhttpclient.InferenceServerClient("localhost:8000")
+        inputs = []
+        inputs.append(tritonhttpclient.InferInput('INPUT', [1], 'INT32'))
+        inputs.append(tritonhttpclient.InferInput('TEST_CASE', [1], 'INT32'))
+        inputs[0].set_data_from_numpy(np.asarray([1], dtype=np.int32))
+        inputs[1].set_data_from_numpy(np.asarray([1], dtype=np.int32))
+        correlation_id = 3
+
+        # Make sure the state is never updated.
+        result_start = triton_client.infer(model_name="no_state_update", inputs=inputs, sequence_id=correlation_id, sequence_start=True)
+        self.assertEqual(result_start.as_numpy('OUTPUT')[0], 1)
+        for _ in range(10):
+            result = triton_client.infer(model_name="no_state_update", inputs=inputs, sequence_id=correlation_id)
+            self.assertEqual(result.as_numpy('OUTPUT')[0], 1)
+
+        result_start = triton_client.infer(model_name="no_state_update", inputs=inputs, sequence_id=correlation_id, sequence_end=True)
+        self.assertEqual(result.as_numpy('OUTPUT')[0], 1)
+
+    def test_request_output_not_allowed(self):
+        triton_client = tritonhttpclient.InferenceServerClient("localhost:8000")
+        inputs = []
+        inputs.append(tritonhttpclient.InferInput('INPUT', [1], 'INT32'))
+        inputs[0].set_data_from_numpy(np.asarray([1], dtype=np.int32))
+
+        outputs = []
+        outputs.append(tritonhttpclient.InferRequestedOutput('OUTPUT_STATE'))
+
+        for backend in BACKENDS.split(" "):
+            with self.assertRaises(InferenceServerException) as e:
+                triton_client.infer(
+                    model_name=f"{backend}_nobatch_sequence_int32",
+                    inputs=inputs,
+                    outputs=outputs,
+                    sequence_id=1,
+                    sequence_start=True,
+                    sequence_end=True)
+            self.assertTrue(str(e.exception).startswith("unexpected inference output 'OUTPUT_STATE' for model"))
+
+    def test_request_output(self):
+        triton_client = tritonhttpclient.InferenceServerClient("localhost:8000")
+        inputs = []
+        inputs.append(tritonhttpclient.InferInput('INPUT', [1], 'INT32'))
+        inputs[0].set_data_from_numpy(np.asarray([1], dtype=np.int32))
+
+        outputs = []
+        outputs.append(tritonhttpclient.InferRequestedOutput('OUTPUT_STATE'))
```

Comment:
So different from the previous case,  in this test, the request for the state works because the model config specifies the state in output field as well?

Reply:
Right. The model must specify the state in the output field if it wants to allow the client to request it.